src/main/java/ch/uzh/ifi/hase/soprafs24/service/TaskService.java

```diff
@@ -80,71 +71,106 @@ public Task createTask(Task newTask, long userId) {
         return newTask;
     }
 
-    public void apply(TaskPutDTO taskPutDTO, String token){
-        User candidate = userRepository.findUserByToken(token);
-        //to check if there is a token or the token has been manipulated
-        if (candidate==null || token.isEmpty() || taskPutDTO.getUserId()!=candidate.getId()){
+    @Transactional
+    public void apply(TaskPutDTO taskPutDTO, String token) {
+        User candidate = userService.getUserByToken(token);
+
+        // Check for valid token and user
+        if (candidate == null || token.isEmpty() || taskPutDTO.getUserId() != candidate.getId()) {
             throw new ResponseStatusException(HttpStatus.FORBIDDEN, "Invalid token.");
         }
-        long taskId= taskPutDTO.getTaskId();
 
-        Task selectedTask = taskRepository.findById(taskId);
-        if (selectedTask == null){
-            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "The selected task does not exists.");
-        }
-        Application existingApplication = applicationsRepository.findByUserAndTask(candidate, selectedTask);
-        if (existingApplication != null) {
+        long taskId = taskPutDTO.getTaskId();
+        Task selectedTask = taskRepository.findById(taskId).orElseThrow(() ->
+                new ResponseStatusException(HttpStatus.NOT_FOUND, "The selected task does not exist."));
+
+        // Check if the candidate has already applied
```

Comment:
This replaces the call to the applicationRepository to check if the candidate has already applied